_drafts/best-of-binutils.md

```diff
@@ -0,0 +1,1011 @@
+
+#### Useful Arguments
+
+- `-a` To show the address which was decoded. This can be useful when decoding multiple addresses
+  at once or using `addr2line` in a batch processing mode.
+- `-f` To display the function name the address is in
+- `-e <ELF>` To specify the excutable to decode. (Otherwise, `addr2line` checks for `a.out` in the
+  working directory and uses that.)
+
+#### Example Usage
+
+You can run it on a single address:
+
+```bash
+$ arm-none-eabi-addr2line -a -f -e build/zephyr/zephyr.elf 0x08000946
+  0x08000946
+_printk_dec_ulong
+zephyr/lib/os/printk.c:465
+
+```
+
+or pipe a bunch of addresses in from stdin and get them all decoded at once
+
+```bash
+$ echo -e "0x8000946\n0x8000336" |
+   arm-none-eabi-addr2line -a -f -e build/zephyr/zephyr.elf
+0x08000946
+_printk_dec_ulong
+zephyr/lib/os/printk.c:465
+0x08000336
+__udivmoddi4
+```
+
+### size
+
+`size` walks through the sections in the ELF and sums up the totals for all sections which are
+allocated. This can be useful for getting a quick summary of what the binary footprint looks like:
+
+```bash
+$ arm-none-eabi-size build/zephyr/zephyr.elf
+text data bss dec hex filename
+84340 2416 32157 118913 1d081 build/zephyr/zephyr.elf
+```
+
+## Other useful tools for examining ELFs
+
+There's a number of additional tools aside from GNU Binutils which I've found tremendously useful
+over the years. While a full discussion of them is outside the scope of this article (and merits a
+blog post of their own), I've included some of my favorite highlights here.
+
+### Godbolt (aka Compiler Explorer)
+
+No discussion of binary inspection is complete without a shout out to Compiler Explorer. This is an online tool
+that allows you to compile snippets of code against a ridiculous amount of different compilers. Even better you can
+create short urls and share links with friends.
+
+For example, [here is](https://godbolt.org/z/iqeNUv) an interesting one I was looking at recently.
+
+![](/img/best-of-binutils/compiler-explorer-example.png)
+
+For ARMv7-M, 4 byte loads and store instructions (`ldr` and `str`, respectively) support unaligned
+access. When compiling code for space optimizations (using `-Os` CFLAG), I would have expected unaligned
+accesses to get optimized using this instruction. What I noticed is for some reason if the
+unaligned store is through a packed structure, the optimization does not take place.
+
+### pyelftools
+
+Pyelftools is a pure python package capable of parsing ELF and DWARF information. If you are
+looking to perform your own analysis on ELF files, I'd highly recommend checking out the
+repo. There is an [examples](https://github.com/eliben/pyelftools/tree/master/examples) folder in
+the project that you can use as inspiration to get started.
+
+#### Pure Python GNU Binutils size Command Equivalent
+
+As a quick example, we can implement an equivalent of the `size` binutil command in about 30 lines
+of python!
```

Comment:
```suggestion
of Python!
```

Reply:
oof I was on the struggle bus capitalizing python